### ISSUE TITLE
.from_features() does not accept FeatureCollection

### ISSUE THREAD

**Issue by micahcochran** [geopandas/geopandas#225]
The `__geo_interface__` internally stores FeatureCollections.

However, `Geodataframe.from_features()` does not accept FeatureCollections as input.  [My test of .from_features()](https://gist.github.com/micahcochran/7826e025aa1f0f53e45b).  

What should be the preferred way to get a FeatureCollection from the `__geo_interface__` into geopandas?  Here are the options that I've thought of:
1. The existing `from_feature()` function could be extended to add FeatureCollections.
2. Create a new function `from_featurecollection()`
3. Do nothing and encourage developers to convert a FeatureCollection into a list of features.  Admittedly not that hard.

I have code that I think would be a good starting point.  I would like to get some feedback about the preferred way to do this.

**Comment by perrygeo** [comment]
I typically do option 3, which works ok

    df_cities = gpd.GeoDataFrame.from_features(global_cities.__geo_interface__['features'])

But it option 1 is a much tidier interface; treat a collection as a list of features. Internally, we could check if the input type is `FeatureCollection` and iterate over `features`. 

    df_cities = gpd.GeoDataFrame.from_features(global_cities)

If there were crs info in the collection, we could grab that as well.

**Comment by micahcochran** [comment]
For option 1, I had envisioned keeping `from_features()` to have a list parameter of `Feature`s and `FeatureCollection`s.  This would allow for easier merging of geopanda datasets, but `append()` might  be the more correct way to accomplish that.

First,I tried the underlying `DataFrame.append()` from Pandas, which seem worked until I tried to writing to a file.  I think `append()` needs to be geopanda'ed.  I will try to get to the bottom of that and post a new issue.

**Comment by jorisvandenbossche** [comment]
@micahcochran Why do you need a DataFrame.append? I think you can just detect first if `features` is a list of features or a FeatureCollection, and if it is a FeatureCollection take the list of features inside it, and the rest of the code should work as it is? Or am I missing something?

**Comment by micahcochran** [comment]
@jorisvandenbossche `DataFrame.append()` seemed more intuitive to me than creating a new `GeoDataFrame.from_features()`.  [It is close to what ArcGIS Desktop's Append tool does](http://help.arcgis.com/EN/arcgisdesktop/10.0/help/index.html#//001700000050000000), but is exactly like the [ArcGIS Desktop Merge tool](http://help.arcgis.com/EN/arcgisdesktop/10.0/help/index.html#/Merge/001700000055000000/)--nothing like Pandas merge.

I figured out that I could use the `.append()`, if followed up with a call to `.set_geometry()`, which adds the geometry column and crs metadata to the GeoDataFrame.

I'm using Geopandas for unifying datasets (many datasets become one dataset) and adding attributes in an automated fashion.  I'm using this to automate routine GIS data maintenance for local government, not science or even analysis.

I've figured out that the extra couple lines of code to make this function work isn't that big of a deal.  It is just a learning curve.

**Comment by emiliom** [comment]
> 1. The existing `from_feature()` function could be extended to add FeatureCollections
:+1: 

While doing 
```python
df_cities = gpd.GeoDataFrame.from_features(global_cities.__geo_interface__['features'])
```
is easy once you know about it or if you're familiar with `__geo_interface__`, it's not at all obvious to others.

**Comment by jorisvandenbossche** [comment]
@micahcochran interested doing a PR to improve `from_features` to handle FeatureCollections? (your option 1)

**Comment by emiliom** [comment]
`from_features` method is at https://github.com/geopandas/geopandas/blob/master/geopandas/geodataframe.py#L172 (to help me, if nothing else)

I like @perrygeo's summary of what needs to be done: "treat a collection as a list of features. Internally, we could check if the input type is `FeatureCollection` and iterate over `features`"

**Comment by jorisvandenbossche** [comment]
@emiliom if you want to do a PR, that is certainly also welcome!

**Comment by emiliom** [comment]
> @emiliom if you want to do a PR, that is certainly also welcome!

I'm definitely considering it. Stay tuned.

**Comment by jorisvandenbossche** [comment]
@emiliom Great! If you have any questions, just ask!

**Comment by emiliom** [comment]
I've implemented this enhancement and created [a notebook](https://gist.github.com/emiliom/bde588461f816a0f0b5e4ae47ffbcc20) showing three types of "features" being passed to `from_features`:
1. list of Feature dicts (current functionality)
2. [geojson](http://python-geojson.readthedocs.io) FeatureCollection object (read via its `__geo_interface__`)
3. FeatureCollection dict

I'll submit a PR later, possibly not until tomorrow.

One thing to note is that FeatureCollection attributes/items other than `type` and `features` will not be read. For example, the FeatureCollection `bbox` yielded by `GeoDataFrame.__geo_interface__` will be dropped in a round-trip test from a GDF to a GeoJSON dict and back to GDF. I think that's the conventional expectation for `__geo_interface__`, though; otherwise, where would they go in a GDF?

**Comment by jorisvandenbossche** [comment]
That looks good! 
Yes, I think it is OK to drop the additional fields in the dict outside of properties and geometry. Although 'id' could be set as the index. We do that for converting to features as well (https://github.com/geopandas/geopandas/pull/421)

**Comment by emiliom** [comment]
> Yes, I think it is OK to drop the additional fields in the dict outside of properties and geometry. Although 'id' could be set as the index. We do that for converting to features as well (#421)

In the GeoJSON IETF specs [FeatureCollection objects](https://tools.ietf.org/html/rfc7946#section-3.3) only have (besides the type) a mandated `features` member, and an optional `bbox` member. id's are definitely useful at the Feature level, of course, but an id at the level of the FeatureCollection doesn't seem like a widely used or useful attribute, as it'll have just one value. Where in a geodataframe would we attach a single id or other FeatureCollection-level attributes? Even the optional FeatureCollection-level bbox seems irrelevant, come to think of it: once the features are read in and the geodataframe created, the geodataframe provides the bounding box via its `total_bounds` property.

So I guess I'm leaning towards sticking with my initial assumption: for FeatureCollection, only `type` and `features` should be read. For the already existing functionality (reading a list of features), I'm leaving things the way they are.

**Comment by jorisvandenbossche** [comment]
Sorry, I mean't the 'id' keys inside a feature, not the 'id' key of the FeatureCollection. I agree that the last one should not be used. 
Since we are using the index as the feature ids when outputting (eg in `iterfeatures` or in `__geo_interface__`), I think it makes sense to also read then in `from_features`. But let's leave that for another PR.

**Comment by emiliom** [comment]
Sounds like a good idea, but agreed that it's for another PR, and probably a separate issue.
